pocketimdb/movies/views.py

```diff
@@ -1,14 +1,57 @@
+from django.db.models import Count, Sum, Q, functions, F, Case, When
+from django.db.models.functions import Coalesce
 from rest_framework import viewsets, mixins, filters
+from rest_framework.decorators import action
 from rest_framework.permissions import IsAuthenticated
-from .models import Movie
-from .serializers import MovieSerializer
+from rest_framework.response import Response
+from rest_framework.status import HTTP_200_OK, HTTP_201_CREATED, HTTP_204_NO_CONTENT, HTTP_400_BAD_REQUEST, HTTP_403_FORBIDDEN
+from .models import Movie, MovieLike
+from .serializers import MovieSerializer, AddMovieLikeSerializer
 
 class MovieViewSet(mixins.ListModelMixin,
                 mixins.RetrieveModelMixin,
                 viewsets.GenericViewSet):
 
-    queryset = Movie.objects.all()
     serializer_class = MovieSerializer
     permission_classes = [IsAuthenticated]
     filter_backends = [filters.SearchFilter]
-    search_fields = ['title']
\ No newline at end of file
+    search_fields = ['title']
+
+    def get_queryset(self):
+        return Movie.objects.annotate(
+            likes=Coalesce(Count('movie_likes__like', filter=Q(movie_likes__like=MovieLike.Like.LIKE)), 0),
+            dislikes=Coalesce(Count('movie_likes__like', filter=Q(movie_likes__like=MovieLike.Like.DISLIKE)), 0),
+            user_liked_or_disliked=Coalesce(Sum('movie_likes__like', filter=Q(movie_likes__user=self.request.user)), 0),
+        ).order_by('id')
+
+    @action(detail=True, url_path='likes')
+    def like(self, request, pk):
+        pass
```

Comment:
cemu ovo sluzi ?

Reply:
To pretstavlja custom rutu, ali posto imam vise HTTP metoda za tu rutu, onda na taj nacin definisem osnovu, a ispod sa `@like.mapping.post` npr. dekorisem pojedinacne metode koje za tu rutu obradjuju definisanu HTTP metodu.

Moze ovo i drugacije, ali ovo sam negde video i ucinilo mi se okej da razdvojim to u vise metoda. Drugi nacin bi bio da sve u jednoj metodi imam, pa da imam neki switch u ondosu na HTTP metodu ili tako nesto.

Reply:
I ovo sam promenio na ovaj drugi nacin ipak, jer na ovaj nacin mi je dozvoljavao `GET` zahtev na tu rutu i on ocekuje da to po default-u implementiram u toj osnovnoj metodi.